src/backend/parser/parse_agg.c

```diff
@@ -1505,16 +1536,138 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
 															   targetList));
 		}
 	}
 
 	return result;
 }
 
+static List*
+get_com_grouping_ressortgroupref_routine(Node *grpcl, List *targetList, List *com_refs)
+{
+	List *result = NIL;
+
+	if ( !grpcl )
+		return result;
+
+	Assert(IsA(grpcl, SortGroupClause) ||
+		   IsA(grpcl, GroupingClause) ||
+		   IsA(grpcl, List));
+
+	if (IsA(grpcl, SortGroupClause))
+	{
+		TargetEntry *tle = get_sortgroupclause_tle((SortGroupClause *) grpcl, targetList);
+		result = lappend_int(result, tle->ressortgroupref);
+	}
+	else if (IsA(grpcl, GroupingClause))
+	{
+		ListCell *l;
+		GroupingClause *gc = (GroupingClause*)grpcl;
+
+		/* Cube and Rollup do not contain common attributes */
+		if (gc->groupType == GROUPINGTYPE_CUBE || gc->groupType == GROUPINGTYPE_ROLLUP)
+		{
+			list_free(com_refs);
+			return NIL;
+		}
+
+		foreach(l, gc->groupsets)
+		{
+			List *grouping_refs = get_com_grouping_ressortgroupref_routine((Node*)lfirst(l), targetList, com_refs);
+			
+			/* An empty grouping has no common attributes */
+			if (!grouping_refs)
+				return NIL;
+
+			/*Exclude refs that did not appear in this grouping*/
+			ListCell *lc = list_head(com_refs);
+			while (lc) {
+				if (list_member_int(grouping_refs, lfirst_int(lc)))
+				{
+					lc = lnext(lc);
+				}
+				else
+				{
+					ListCell *lc_del = lc;
+					lc = lnext(lc);
+					com_refs = list_delete_int(com_refs, lfirst_int(lc_del));					
+				}
+			}
+		}
+
+		return com_refs;
+	}
+	else
+	{
+		List *tles = (List *)grpcl;
+		ListCell *lc;
+
+		foreach(lc, tles)
+		{
+			List *chunk = get_com_grouping_ressortgroupref_routine((Node *)lfirst(lc), targetList, com_refs);
+			if(!chunk)
+				return NIL;
+
+			result = list_concat(result, chunk);
+		}
+	}
+
+	return result;
+}
+
+/*
+ * get_com_grouping_tles -
+ *     Return a list of common ressortgroupref expressions appeared in group
+ * 	   clauses.
+ */
+List*
+get_com_grouping_ressortgroupref(Query *qry, List *grp_tles){
+	if (!qry || !grp_tles)
+		return NIL;
+
+	List *com_grouping_ressortgroupref = NIL;
+	ListCell *gc;
+
+	foreach(gc, qry->groupClause)
+	{
+		List *com_group_expr_ressortgroupref = NIL;
+		Node *grp = lfirst(gc);
+
+		if (grp == NULL)
+			continue;
+
+		/* Scan in grouping expression*/
+		if (IsA(grp, GroupingClause))
+		{
+			/* We assume that attributes are present in current grouping expression. */
+			ListCell *lc;
+			foreach(lc, grp_tles)
+			{
+				TargetEntry* te = (TargetEntry*) lfirst(lc);
+				com_group_expr_ressortgroupref =
+					list_append_unique_int(com_group_expr_ressortgroupref, te->ressortgroupref);
+			}
+			com_group_expr_ressortgroupref =
+				get_com_grouping_ressortgroupref_routine(grp, qry->targetList,com_group_expr_ressortgroupref);
+		}
+		/* Scan in group by*/
+		if (IsA(grp, SortGroupClause))
+		{
+			TargetEntry *tle = get_sortgroupclause_tle((SortGroupClause *) grp, qry->targetList);
+			com_group_expr_ressortgroupref = lappend_int(com_group_expr_ressortgroupref, tle->ressortgroupref);
```

Comment:
```suggestion
			com_grouping_ressortgroupref = list_append_unique_int(com_grouping_ressortgroupref, tle->ressortgroupref);
```

Thus
```
List *com_group_expr_ressortgroupref = NIL;
``` 
and 
```
/* Form a list of common refs for all clauses */
com_grouping_ressortgroupref = list_concat_unique_int(com_grouping_ressortgroupref, com_group_expr_ressortgroupref);
list_free(com_group_expr_ressortgroupref);
```
can be moved inside the block
```
if (IsA(grp, GroupingClause))
{
}
```

Reply:
Fixed